pkgs/development/compilers/c3c/default.nix

```diff
@@ -37,6 +46,15 @@ llvmPackages.stdenv.mkDerivation rec {
     cp -r lib $out
   '';
 
+  patches = [
+    ./add_ldd_link.patch
```

Comment:
Normally we would fetch the patch from upstream if possible, and also add an explanatory comment

Reply:
This patch is sort of specific to nix as to have a way of finding the LLVM lld libs. I will try to push this upstream, but in the meantime I think this can stay here.

Will add an explanatory comment as well.

Reply:
Update: Took a dive and researched this more as I thought that there might be something more here. 

The question is whether or not we actually need the patch/modify the CMakeLists regarding the lib paths and why.

On the upstream side, the intent is that only the libs that are compiled with a certain llvm version are supposed to be used. i.e. use only the lib*lld.so compiled with and for use for llvm 16. This is a valid thing to do, rather than rely on CMake finding them from its search path.

On Nix side, in the LLVM_LIBRARY_DIRS cmake variable you only have `/nix/store/...-llvm/lib` dir, which is valid from a Nix standpoint. Distros such as Ubuntu/Debian, put the libs in the llvm lib dir (i.e. [Debian liblld14 filelist](https://packages.debian.org/sid/amd64/liblld-14/filelist) ), because LLVM's cmake file makes use of [GNUInstallDirs](https://cmake.org/cmake/help/latest/module/GNUInstallDirs.html) and it puts them inside `/usr/lib/lib${LLVM_TARGET_PREFIX}`. On Nix side, we put them inside  `${CMAKE_INSTALL_LIB_DIR}/${LLVM_TARGET_PREFIX}` and each llvm project (such as lld) is patched as such which makes LLVM_LIBRARY_DIRS inadequate if you wish to target for example a lld lib.

It does seem to be fixable in an elegant way in LLVM by making `find_package(LLD)` set a `LLD_LIBRARY_DIR`, for example.

So all in all, we need to modify the CMakeLists file for now. Replaced with a substituteInPlace for brevity. Upstream, I will make a PR so that we can have a CMake option, but in the meantime this should be ok.